src/crypto/ring/digest.rs

```diff
@@ -2,44 +2,37 @@
 // License, v. 2.0. If a copy of the MPL was not distributed with this
 // file, You can obtain one at https://mozilla.org/MPL/2.0/.
 
-use core::cmp::PartialEq;
+use core::convert::{TryFrom, TryInto};
+use crypto::digest::Digest;
+use ring_compat::digest::{Sha256, Sha384, Sha512};
 
-use crate::crypto::digest::Digest;
+use crate::crypto::digest::Digests;
+use crate::impl_digests;
 use crate::msgs::algorithms::BaseHashAlgo;
 
-// TODO: Should this even exist? Should a single crypto provider be used for all
-// mechanisms? Should we allow more than one provider per crypto mechanism?
-pub type DigestImpl = RingDigest;
-
-/// A Ring based implementation of a Digest
-#[derive(Debug, Clone)]
-pub struct RingDigest {
-    digest: ring::digest::Digest,
-}
-
-impl PartialEq for RingDigest {
-    fn eq(&self, other: &RingDigest) -> bool {
-        self.as_ref() == other.as_ref()
+impl_digests!(
+    pub enum ProvidedDigests {
+        Sha256(Sha256),
+        Sha384(Sha384),
+        Sha512(Sha512),
     }
-}
+);
 
-impl Eq for RingDigest {}
-
-impl Digest for RingDigest {
-    fn hash(algorithm: BaseHashAlgo, buf: &[u8]) -> Self {
-        let algo = match algorithm {
-            BaseHashAlgo::SHA_256 => &ring::digest::SHA256,
-            BaseHashAlgo::SHA_384 => &ring::digest::SHA384,
-            BaseHashAlgo::SHA_512 => &ring::digest::SHA512,
-            _ => unimplemented!(),
-        };
-        RingDigest { digest: ring::digest::digest(algo, buf) }
+impl ProvidedDigests {
```

Comment:
This should almost certainly be part of a trait that we can constrain `Digests` on. That way implementers of `ProvidedDigests` enums will get good error messages to remind them to implement this.